kona/key_value_store_lmdb.py

```diff
@@ -0,0 +1,197 @@
+
+    @_validate_args_bytes_without_first
+    @_error_convert
+    def delete(self, key: bytes):
+        self._txn.delete(key)
+
+    @_error_convert
+    def clear(self):
+        self._txn.abort()
+
+    @_error_convert
+    def write(self):
+        self._txn.commit()
+
+
+class _KeyValueStoreCancelableWriteBatchLMDB(KeyValueStoreCancelableWriteBatch):
+    def __init__(self, store: KeyValueStore, db: lmdb.Environment):
+        super().__init__(store)
+        self._original_items = dict()
+        self._db = db
+
+    def _touch(self, key: bytes):
+        if key in self._original_items:
+            return
+
+        value = self._db.begin().get(key, None)
+        self._original_items[key] = value
+
+    def _get_original_touched_item(self):
+        for key, value in self._original_items.items():
+            yield key, value
+
+    def clear(self):
+        super().clear()
+        self._original_items.clear()
+
+    def close(self):
+        self._original_items: dict = None
+
+
+class KeyValueStoreLMDB(KeyValueStore):
+    def __init__(self, uri: str, **kwargs):
+        uri_obj = urllib.parse.urlparse(uri)
+        if uri_obj.scheme != 'file':
+            raise ValueError(f"Support file path URI only (ex. file:///xxx/xxx). uri={uri}")
+        self._path = f"{(uri_obj.netloc if uri_obj.netloc else '')}{uri_obj.path}"
+        self._db = self._new_db(self._path)
+
+    @_error_convert
+    def _new_db(self, path) -> lmdb.Environment:
+        return lmdb.Environment(path)
+
+    @_validate_args_bytes_without_first
+    @_error_convert
+    def get(self, key: bytes, *, default=None, **kwargs) -> bytes:
+        if default is not None:
+            _validate_args_bytes(default)
+
+        with self._db.begin() as txn:
+            result = txn.get(key, default)
+            if result is None:
+                raise KeyError(f"Has no value of key({key})")
+            return result
+
+    @_validate_args_bytes_without_first
+    @_error_convert
+    def put(self, key: bytes, value: bytes, *, sync=False, **kwargs):
+        with self._db.begin(write=True) as txn:
+            txn.put(key, value)
+
+    @_validate_args_bytes_without_first
+    @_error_convert
+    def delete(self, key: bytes, *, sync=False, **kwargs):
+        with self._db.begin(write=True) as txn:
+            txn.delete(key)
+
+    @_error_convert
+    def close(self):
+        if self._db:
+            self._db.close()
+            gc.collect()
+            self._db = None
+
+    @_error_convert
+    def destroy_store(self):
+        self.close()
+
+        def rm_tree(path: Path):
+            for child in path.iterdir():
+                if child.is_file():
+                    child.unlink()
+                else:
+                    rm_tree(child)
+            path.rmdir()
+
+        rm_tree(Path(self._path))
+
+    @_validate_args_bytes_without_first
+    @_error_convert
+    def key_may_exist(self, key: bytes) -> Tuple[bool, Any]:
```

Comment:
I think it's better return `NotImplemented` than `pass`, and also same others methods in children classes of same parent with this if it's not implemented.  Maybe the author of this method missed. 😅